### PR TITLE
Implement OpenLineage spark-extension-interfaces

### PR DESCRIPTION
**Pull request by ddebowczyk92**
This PR adds output table-level lineage by implementing the `LineageRelationProvider` interface. Additionally, it removes the implementation of the `com.google.cloud.spark.bigquery.BigQueryRelationProvider` class, which was a temporary workaround for providing table lineage.

### CODE REVIEW

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/SparkBQLineageProvider.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.spark.bigquery;
+
+import io.openlineage.spark.extension.OpenLineageExtensionProvider;
+
+public class SparkBQLineageProvider implements OpenLineageExtensionProvider {
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/SparkBQLineageProvider.java:20]: Please rename SparkBQLineageProvider -> SparkBigQueryLineageProvider

File: spark-bigquery-connector-common/src/main/resources/META-INF/services/io.openlineage.spark.extension.OpenLineageExtensionProvider [hunk 2]
```diff
@@ -0,0 +1 @@
+com.google.cloud.spark.bigquery.SparkBQLineageProvider
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/resources/META-INF/services/io.openlineage.spark.extension.OpenLineageExtensionProvider:1]: Fixed to renamed class

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml [hunk 3]
```diff
@@ -27,6 +27,11 @@
             <artifactId>spark-bigquery-connector-common</artifactId>
             <version>${project.parent.version}</version>
         </dependency>
+        <dependency>
+            <groupId>io.openlineage</groupId>
+            <artifactId>spark-extension-interfaces</artifactId>
+            <version>1.19.0</version>
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml:33]: The version should only appear in the spark-bigquery-parent POM file, as part of the dependencyManagement

File: spark-bigquery-dsv2/spark-bigquery-dsv2-parent/pom.xml [hunk 4]
```diff
@@ -37,6 +37,8 @@
     <dependency>
       <groupId>io.openlineage</groupId>
       <artifactId>openlineage-spark_${scala.binary.version}</artifactId>
+      <version>1.19.0</version>
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-dsv2-parent/pom.xml:40]: The version should only appear in the spark-bigquery-parent POM file, as part of the dependencyManagement

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/TestSparkBQLineageExtensionProvider.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+public class TestSparkBQLineageExtensionProvider implements OpenLineageExtensionProvider {
+  @Override
+  public String shadedPackage() {
+    return "com.google.cloud.spark.bigquery.repackaged.io.openlineage.spark.shade";
+  }
+
+  @Override
+  public String getVisitorClassName() {
+    return "io.openlineage.spark.shade.extension.v1.lifecycle.plan.SparkOpenLineageExtensionVisitor";
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/TestSparkBQLineageExtensionProvider.java:28]: Wouldn't this class be shaded? shouldn't it be `com.google.cloud.spark.bigquery.repackaged.io.openlineage.spark.shade.extension.v1.lifecycle.plan.SparkOpenLineageExtensionVisitor`?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/SparkBQLineageProvider.java (hunk 1 above)
Comment (ddebowczyk92) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/SparkBQLineageProvider.java]: Done

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml (hunk 3 above)
Comment (ddebowczyk92) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml]: Done

File: spark-bigquery-dsv2/spark-bigquery-dsv2-parent/pom.xml (hunk 4 above)
Comment (ddebowczyk92) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-parent/pom.xml]: Done

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/TestSparkBQLineageExtensionProvider.java (hunk 5 above)
Comment (ddebowczyk92) [reply spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/TestSparkBQLineageExtensionProvider.java:28]: The issue is that, for tests, classes provided to the classpath are not shaded. Therefore, we need an alternative `OpenLineageExtensionProvider` that will point to the unshaded resource. Hence, we have introduced the `spark.openlineage.testExtensionProvider` config property to ensure that only one `OpenLineageExtensionProvider` is selected.

Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/TestSparkBQLineageExtensionProvider.java:28]: Your can return `SparkOpenLineageExtensionVisitor.class.getCanonicalName()`, the import will change as part of the shading.

Comment (ddebowczyk92) [reply spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/TestSparkBQLineageExtensionProvider.java]: you are right!

**Comment by ddebowczyk92** [comment]
Hey @davidrabinowitz, the build failed last time because I removed openlineage-spark dependency from `spark-bigquery_2.13` pom, now it's restored. I tested it locally and everything passed through, could you retry the build?

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**